### PR TITLE
[SYCL] Fix memory leak in SYCL context after expected exception

### PR DESCRIPTION
**Pull request by dm-vodopyanov**
This patch fixes memory leak in SYCL context which happened
after expected exception

### CODE REVIEW

**Comment by dm-vodopyanov** [comment]
Test: https://github.com/intel/llvm-test-suite/pull/280

File: sycl/source/detail/scheduler/scheduler.cpp [hunk 1]
```diff
@@ -108,11 +108,15 @@ EventImplPtr Scheduler::addCG(std::unique_ptr<detail::CG> CommandGroup,
   {
     std::shared_lock<std::shared_timed_mutex> Lock(MGraphLock);
 
-    Command *NewCmd = static_cast<Command *>(NewEvent->getCommand());
+    // enqueueCommand() func below may throw an exception, so use shared_ptr
+    // here to avoid memory leak
+    std::shared_ptr<Command> NewCmd(
```
Comment (s-kanaev) [inline sycl/source/detail/scheduler/scheduler.cpp:113]: The smart ptr is owner of the memory of command now. It'll free the memory upon destructor call.
Deallocation is only performed in some corner case here.
Hence, a call to `release()` (which isn't present for `shared_ptr`) method is required in the end in general case and call to `reset()` in the corner case.

Comment (s-kanaev) [inline sycl/source/detail/scheduler/scheduler.cpp:113]: Is it exclusively `shared_ptr` which is needed here? Will `unique_ptr` fulfill the needs?

**Review by s-kanaev — CHANGES_REQUESTED** [review]

File: sycl/source/detail/scheduler/scheduler.cpp (hunk 1 above)
Comment (dm-vodopyanov) [reply sycl/source/detail/scheduler/scheduler.cpp]: Yes, `shared_ptr` doesn't need at all, replaced with `unique_ptr`.

Comment (dm-vodopyanov) [reply sycl/source/detail/scheduler/scheduler.cpp]: BTW, corner case is the case when exception is thrown from `GraphProcessor::enqueueCommand()` function, and stack starts to unwind. I think there is no place to add a call to `reset()` function in such case. What do you think?

File: sycl/source/detail/scheduler/scheduler.cpp [hunk 2]
```diff
@@ -108,26 +108,30 @@ EventImplPtr Scheduler::addCG(std::unique_ptr<detail::CG> CommandGroup,
...
       // resources may be in use by backend and synchronization point here is
       // at native kernel execution finish.
-      if (NewCmd->MDeps.size() == 0 && NewCmd->MUsers.size() == 0) {
+      if (NewCmd.get()->MDeps.size() == 0 && NewCmd.get()->MUsers.size() == 0) {
         if (IsHostKernel)
-          static_cast<ExecCGCommand *>(NewCmd)->releaseCG();
+          static_cast<ExecCGCommand *>(NewCmd.get())->releaseCG();
 
         NewEvent->setCommand(nullptr);
-        delete NewCmd;
```
Comment (s-kanaev) [inline sycl/source/detail/scheduler/scheduler.cpp:128]: Why is this line removed? Should it be replaced with `unique_ptr::reset()` to deallocate the command? I believe, it should as the command isn't tracked anymore.

File: sycl/source/detail/scheduler/scheduler.cpp (hunk 1 above)
Comment (s-kanaev) [reply sycl/source/detail/scheduler/scheduler.cpp]: The corner case I was talking about is the `delete NewCmd`, which was removed: https://github.com/intel/llvm/pull/3749/files#diff-583d91db0b0dd79af9e56cdc4550fe76ee97271aa73955241b7a316907f63c5eL128

As for the exception thrown by `GraphProcessor::enqueueCommand()`, is it an option to catch the exception?

Comment (dm-vodopyanov) [reply sycl/source/detail/scheduler/scheduler.cpp]: Fixed.

> As for the exception thrown by `GraphProcessor::enqueueCommand()`, is it an option to catch the exception?

Such async exception should be caught on the user's side using async exception handler passed to `queue`.

File: sycl/source/detail/scheduler/scheduler.cpp (hunk 2 above)
Comment (dm-vodopyanov) [reply sycl/source/detail/scheduler/scheduler.cpp:128]: Fixed.

**Review by s-kanaev — APPROVED** [review]
LGTM

**Comment by bader** [comment]
@dm-vodopyanov, please, fix regressions.

**Review by s-kanaev — APPROVED** [review]
LGTM
